### DISCUSSION TITLE
Fast(er) multiplications, divisions, square roots?

### DISCUSSION THREAD

**Discussion by mskyttner** [duckdb/duckdb#2695] [Ideas]
This news story at https://newsroom.unsw.edu.au/news/science-tech/maths-whiz-solves-48-year-old-multiplication-problem and the paper "Integer multiplication in time O(n log n)" at https://hal.archives-ouvertes.fr/hal-02070778/document says there is a fast(er) way to compute integer multiplication with applications for division, square roots etc:

> It means you can do all sorts of arithmetic more efficiently, for example division and square roots. You could also calculate digits of pi more efficiently than before. It even has applications to problems involving huge prime numbers. People have been hunting for such an algorithm for almost 50 years....

Maybe duckdb could make use of these findings to provide even faster calculations for certain aggregates?

I don't know if there is a C++ implementation available already (and couldn't find any in a quick search), but perhaps some wizards around here would know better whether it would be "worth it" to look into or not?

**Comment by l1t1** [comment]
https://www.zhihu.com/tardis/sogou/qus/317392704

**Comment by hannes** [comment]
This seems to be about very very large numbers "But for numbers with enough digits – billion, trillions or even gazillions – the new algorithm, [...], would outrun even Schönhage and Strassen’s algorithm." In the end, we use whatever multiplication is available in the CPU and don't really have a choice. So I don't think this is very applicable to DuckDB.

**Reply by mskyttner** [reply]
Thanks! What about faster modulo arithmetics in duckdb or would that also not be applicable? "Faster Remainder by Direct Computation: Applications to Compilers and Software Libraries" with a paper at https://arxiv.org/abs/1902.01961 and an implementation in C https://github.com/lemire/fastmod and in go https://github.com/bmkessler/fastdiv. The author is also involved in the simdjson library.
